### ISSUE TITLE
UP047 Fails to indentify functions that have `Callable` in their signature

### ISSUE THREAD

**Issue by Kroppeb** [astral-sh/ruff#19863]
### Summary

I noticed something weird with UP047, `foo` triggers a warning while `foo2` does not.

```python
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")


def foo(t: list[T]) -> T:
    return t[0]


def foo2(t: list[T], c: Callable[[T], None]) -> T:
    c(t[0])
    return t[1]
```

Playground: https://play.ruff.rs/cd7af9a8-0532-40fe-83c7-4282a658808c

### Version

v0.12.7

**Comment by MichaReiser** [comment]
Thanks for opening an issue. This was also discussed in our Discord ([conversation](https://discord.com/channels/1039017663004942429/1070132471699607623/1402996211404902531)). 


Specifically, @AlexWaygood replied

> There was a bug in collections.abc.Callable on Python 3.9 IIRC, which means it isn't safe to replace typing.Callable with collections.abc.Callable on all Python versions (it works most of the time on py39, but not in some edge cases, I think?) 
Should be safe on the newest Python versions, though. Can't remember exactly when the bug was fixed

**Comment by ntBre** [comment]
I talked to Alex, and I don't think that Python bug should affect this rule, which is only replacing the type of generics, not importing `Callable` from a different module. I believe he was thinking of [deprecated-import (UP035)](https://docs.astral.sh/ruff/rules/deprecated-import/#deprecated-import-up035), which will replace `typing.Callable` with `collections.abc.Callable` after Python 3.10.

**Comment by AlexWaygood** [comment]
Yup, @ntBre is correct — my bad! sorry about that 🫠
